sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusProcessorClient.java

```diff
@@ -310,85 +311,105 @@ private synchronized void receiveMessages() {
+                } else {
+                    Context processSpanContext = null;
+                    try {
+                        ServiceBusReceivedMessageContext serviceBusReceivedMessageContext =
+                            new ServiceBusReceivedMessageContext(receiverClient, serviceBusMessageContext);
+
+                        processSpanContext =
+                            startProcessTracingSpan(serviceBusMessageContext.getMessage(),
+                                receiverClient.getEntityPath(), receiverClient.getFullyQualifiedNamespace());
+                        if (processSpanContext.getData(SPAN_CONTEXT_KEY).isPresent()) {
+                            serviceBusMessageContext.getMessage().addContext(SPAN_CONTEXT_KEY, processSpanContext);
+                        }
+                        processMessage.accept(serviceBusReceivedMessageContext);
+                        endProcessTracingSpan(processSpanContext, Signal.complete());
+                    } catch (Exception ex) {
+                        handleError(new ServiceBusException(ex, ServiceBusErrorSource.USER_CALLBACK));
+                        endProcessTracingSpan(processSpanContext, Signal.error(ex));
+                        if (!processorOptions.isDisableAutoComplete()) {
+                            LOGGER.warning("Error when processing message. Abandoning message.", ex);
+                            abandonMessage(serviceBusMessageContext, receiverClient);
                         }
-                    }
-                    if (isRunning.get()) {
-                        LOGGER.verbose("Requesting 1 more message from upstream");
-                        subscription.request(1);
                     }
                 }
+                if (isRunning.get()) {
+                    LOGGER.verbose("Requesting 1 more message from upstream");
+                    subscription.request(1);
+                }
+            }
 
-                @Override
-                public void onError(Throwable throwable) {
-                    LOGGER.info("Error receiving messages.", throwable);
-                    handleError(throwable);
-                    if (isRunning.get()) {
-                        restartMessageReceiver(subscription);
-                    }
+            @Override
+            public void onError(Throwable throwable) {
+                LOGGER.info("Error receiving messages.", throwable);
+                handleError(throwable);
+                if (isRunning.get()) {
+                    restartMessageReceiver(subscription);
                 }
+            }
 
-                @Override
-                public void onComplete() {
+            @Override
+            public void onComplete() {
+                // For rolling session receiver, when one session is disposed by idle, it will emit complete signal,
+                // and we don't want to restart message receiver
+                if (!receiverOptions.isRollingSessionReceiver()) {
```

Comment:
We didn't have this check before, does it mean that the receiver will restart in rolling session receiver?

Reply:
This check is to exclude the rolling session receiver. Because one session may become idle and emit the complete signal, we don't want to restart receiver as we are still receiving from other sessions. Previously, we use `Flux.merge()` to merge all sessions, the subscriber behavior is different. A longing running test is needed to verify this change, I will test that.